ql/src/java/org/apache/hadoop/hive/ql/DriverState.java

```diff
@@ -86,16 +86,35 @@ public void compiling() {
     driverState = State.COMPILING;
   }
 
+  public void compilingWithLocking() {
+    lock();
+    try {
+      driverState = State.COMPILING;
+    } finally {
+      unlock();
+    }
+  }
+
   public boolean isCompiling() {
     return driverState == State.COMPILING;
   }
 
-  public void compilationInterrupted(boolean deferClose) {
-    driverState = deferClose ? State.EXECUTING : State.ERROR;
+  public void compilationInterruptedWithLocking(boolean deferClose) {
```

Comment:
I don't really see these methods really usefull
how about a single method named "transition(newState)"
then the control logic: "wasError" / "deferClose"  can be left outside